examples/matching_engine/src/contract.rs

```diff
@@ -0,0 +1,608 @@
+        &mut self,
+        _context: &CalleeContext,
+        argument: ApplicationCall,
+        _sessions: Vec<SessionId>,
+    ) -> Result<ApplicationCallResult<Self::Message, Self::Response, Self::SessionState>, Self::Error>
+    {
+        let mut result = ApplicationCallResult::default();
+        match argument {
+            ApplicationCall::Order { order } => {
+                self.execute_order_remote(&mut result.execution_result, order)
+                    .await?;
+            }
+        }
+        Ok(result)
+    }
+
+    async fn handle_session_call(
+        &mut self,
+        _context: &CalleeContext,
+        _state: Self::SessionState,
+        _call: (),
+        _forwarded_sessions: Vec<SessionId>,
+    ) -> Result<SessionCallResult<Self::Message, Self::Response, Self::SessionState>, Self::Error>
+    {
+        Err(Self::Error::SessionsNotSupported)
+    }
+}
+
+impl MatchingEngine {
+    fn fungible_id(token_idx: u32) -> Result<ApplicationId<FungibleTokenAbi>, MatchingEngineError> {
+        let parameter = Self::parameters()?;
+        Ok(parameter.tokens[token_idx as usize])
+    }
+
+    /// Calls into the Fungible Token application to receive tokens from the given account.
+    async fn receive_from_account(
+        &mut self,
+        owner: &AccountOwner,
+        count: &Amount,
+        nature: &OrderNature,
+        price: &Price,
+    ) -> Result<(), MatchingEngineError> {
+        let account = Account {
+            chain_id: system_api::current_chain_id(),
+            owner: AccountOwner::Application(system_api::current_application_id()),
+        };
+        let destination = Destination::Account(account);
+        let (amount, token_idx) = Self::get_amount_idx(nature, price, count);
+        self.transfer(*owner, amount, destination, token_idx).await
+    }
+
+    /// Transfers `amount` tokens from the funds in custody to the `destination`.
+    async fn send_to(&mut self, transfer: Transfer) -> Result<(), MatchingEngineError> {
+        let account = Account {
+            chain_id: system_api::current_chain_id(),
+            owner: transfer.owner,
+        };
+        let destination = Destination::Account(account);
+        let owner_app = AccountOwner::Application(system_api::current_application_id());
+        self.transfer(owner_app, transfer.amount, destination, transfer.token_idx)
+            .await
+    }
+
+    async fn transfer(
+        &mut self,
+        owner: AccountOwner,
+        amount: Amount,
+        destination: Destination,
+        token_idx: u32,
+    ) -> Result<(), MatchingEngineError> {
+        let transfer = fungible::ApplicationCall::Transfer {
+            owner,
+            amount,
+            destination,
+        };
+        let token = Self::fungible_id(token_idx).expect("failed to get the token");
+        self.call_application(true, token, &transfer, vec![])
+            .await?;
+        Ok(())
+    }
+
+    //
+    // The order stuff
```

Comment:
No need for this kind of separation in the code. Plus "stuff" is not very informative.

Reply:
ok.